### ISSUE TITLE
Logical discrepancy between transpose function of numpy and pytorch.

### ISSUE THREAD

**Issue by adityamwagh** [pytorch/pytorch#153785]
### 🐛 Describe the bug

There seems to be a discrepancy in the way transpose works in numpy and pytorch. 

```python
>>> import numpy as np
>>> import torch
>>> x = torch.arange(6).reshape(2, 3)
>>> y = x.numpy()
>>> print(x.transpose(0, 1).tolist())
[[0, 3], [1, 4], [2, 5]]
>>> print(y.transpose(0, 1).tolist())
[[0, 1, 2], [3, 4, 5]]
>>> print(y.transpose(1, 0).tolist())
[[0, 3], [1, 4], [2, 5]]
```

Shouldn't `y.transpose(0, 1).tolist()` be equal to `x.transpose(0, 1).tolist()`, where `y` is a pytorch tensor and `x` is that tensor converted to ndarray?

Reference:
[0]: [https://x.com/miru_why/status/1922652289484857549](https://x.com/miru_why/status/1922652289484857549)

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Fedora Linux 42 (Workstation Edition) (x86_64)
GCC version: (Homebrew GCC 11.5.0) 11.5.0
Clang version: 14.0.6
CMake version: version 3.31.6
Libc version: glibc-2.41

Python version: 3.11.11 (main, Dec 19 2024, 14:33:27) [Clang 18.1.8 ] (64-bit runtime)
Python platform: Linux-6.14.6-300.fc42.x86_64-x86_64-with-glibc2.41
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3050 Laptop GPU
Nvidia driver version: 570.144
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12700H
CPU family:                           6
Model:                                154
Thread(s) per core:                   2
Core(s) per socket:                   14
Socket(s):                            1
Stepping:                             3
CPU(s) scaling MHz:                   29%
CPU max MHz:                          2300.0000
CPU min MHz:                          400.0000
BogoMIPS:                             5376.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            544 KiB (14 instances)
L1i cache:                            704 KiB (14 instances)
L2 cache:                             11.5 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Ghostwrite:             Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by fatwir** [comment]
This isn’t a bug but an API-semantics mismatch between PyTorch and NumPy:

1. NumPy’s `ndarray.transpose(*axes)`  
   - If you **pass** axes you must list _all_ dimensions.  
     For a 2×3 array, `y.transpose(0,1)` means “output axis 0 ← old axis 0, output axis 1 ← old axis 1,” i.e. no change.  
   - The **no-arg** form, `y.transpose()` (or `y.T`), defaults to reversing axes → the actual 2D transpose (shape 3×2).

2. PyTorch’s `Tensor.transpose(dim0, dim1)`  
   - Always swaps exactly those two dimensions.  
   - For a 2×3 tensor, `x.transpose(0,1)` gives you the expected 3×2 result.

If you want NumPy’s equivalent of `x.transpose(0,1)`, you can use:  
```python
y.transpose(1,0)
```  
This will produce the same 3×2 output that PyTorch’s `x.transpose(0,1)` gives!

**Comment by adityamwagh** [comment]
Thanks for the clarification. Closing this issue.
